two_factor/views/utils.py

```diff
@@ -194,3 +202,72 @@ def simple_decorator(View):
         View.dispatch = method_decorator(function_decorator)(View.dispatch)
         return View
     return simple_decorator
+
+
+remember_device_cookie_separator = ':'
+
+
+def get_remember_device_cookie(user, otp_device_id):
+    """
+    Compile a signed cookie from user.pk, user.password and otp_device_id,
+    but only return the hashed and signatures and omit the data.
+
+    The cookie is composed of 3 parts:
+    1. A timestamp of signing.
+    2. A hashed value of otp_device_id and the timestamp.
+    3. A hashed value of user.pk, user.password, otp_device_id and the timestamp.
+    """
+    timestamp = baseconv.base62.encode(int(time.time()))
+    cookie_key = hash_remember_device_cookie_key(otp_device_id)
+    cookie_value = hash_remember_device_cookie_value(otp_device_id, user, timestamp)
+
+    cookie = remember_device_cookie_separator.join([timestamp, cookie_key, cookie_value])
+    return cookie
+
+
+def validate_remember_device_cookie(cookie, user, otp_device_id):
+    """
+    Returns True if the cookie was returned by get_remember_device_cookie using the same
+    user.pk, user.password and otp_device_id. Moreover the cookie must not be expired.
+    Returns False if the otp_device_id does not match.
+    Otherwise raises an exception.
+    """
+
+    timestamp, input_cookie_key, input_cookie_value = cookie.split(remember_device_cookie_separator, 3)
+
+    cookie_key = hash_remember_device_cookie_key(otp_device_id)
+    if input_cookie_key != cookie_key:
+        return False
+
+    cookie_value = hash_remember_device_cookie_value(otp_device_id, user, timestamp)
+    if input_cookie_value != cookie_value:
+        raise BadSignature('Signature does not match')
+
+    timestamp_int = baseconv.base62.decode(timestamp)
+    age = time.time() - timestamp_int
+    if age > settings.TWO_FACTOR_REMEMBER_COOKIE_AGE:
+        raise SignatureExpired(
+            'Signature age %s > %s seconds' % (age, settings.TWO_FACTOR_REMEMBER_COOKIE_AGE)
+        )
+
+    return True
+
+
+def hash_remember_device_cookie_key(otp_device_id):
+    return str(base64.b64encode(force_bytes(otp_device_id)))
+
+
+def hash_remember_device_cookie_value(otp_device_id, user, timestamp):
+    salt = 'two_factor.views.utils.hash_remember_device_cookie_value'
+    value = otp_device_id + str(user.pk) + str(user.password) + timestamp
+    return salted_hmac_sha256(salt, value).hexdigest()
+
+
+# inspired by django.utils.crypto.salted_hmac django versions > 3.1a1
+def salted_hmac_sha256(key_salt, value, secret=None):
+    if secret is None:
```

Comment:
No coverage for this line, it just needs a simple test to make sure `salted_hmac_sha256("blah", "blah")` returns a different value from `salted_hmac_sha256("blah", "blah", "my-new-secret")`